Restructured the operators source files and added support for multiple new operators.

eaplatanios opened the pull request:
This PR does some restructuring of the files where ops are defined and also adds the following:

- **Constructors:** `Tensor(stacking:alongAxis:)`, `Tensor(concatenating:alongAxis:)`, and `Tensor(tiling:multiples:)`, along with their VJPs. Also adds `Tensor(zerosLike:)`, `Tensor(onesLike:)`, and `Tensor(rangeFrom:to:stride:)` with tensor-valued arguments.
- **Ops:** Adds `Tensor.unstack()`, `Tensor.split(numSplits:alongAxis:)`, `Tensor.split(sizes:alongAxis:)`, `Tensor.gathering(atIndices:alongAxis:)`,`sigmoid(_:)`, `logSigmoid(_:)`, and `softplus(_:)` along with their VJPs. Also adds `Tensor.batchGathering(atIndices:alongAxis:numBatchDims:)`, `Tensor.gathering(where:alongAxis:)`, and `Tensor.nonZeroIndices()`.
- **Protocols:** Adds an `Optimizable` protocol that makes the optimizers more generally applicable, other than just for layers.

Current issues:

- Cannot backprop through `precondition` which prevents me from making some of the new functions differentiable.

~~This is a work-in-progress PR that also depends on the [advanced indexing PR](https://github.com/apple/swift/pull/23684), but I want to start a discussion and gather feedback about the source files restructuring and also some missing op support. I actually have the following main questions:~~

~~- Is the restructuring of the operators source files acceptable? I believe it will be hard to have all operators in a single source file.~~
~~- I currently cannot add support for gradients of `stacked` and `concatenated` with multiple tensors because I see that `#tfop` does not support ops that output lists of tensors. Are there any updates on this or any current workarounds to return arrays of tensors?~~
~~- Is it ok to merge PRs that add support for new ops but not VJPs yet? I plan to add those to incrementally.~~

~~This PR so far adds support for the `stacked`, `concatenated`, `gathered`, `batchGathered`, `masked`, and `nonZeroIndices` ops, as well as for a new tensor initializer from ranges defined over tensors, and new `stacking`, `concatenating`, and `tiling` tensor initializers.~~

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,350 @@
+    @differentiable(
+        wrt: (x, y),
+        vjp: _vjpSelecting(ifTrue:else:) where T: TensorFlowFloatingPoint)
+    func selecting<T: TensorFlowScalar>(
```
We already have this op. It's called [`replacing(with:where:)`](https://github.com/apple/swift/blob/9340e362563b19eac7a1708c176c97ae900dfa2f/stdlib/public/TensorFlow/Ops.swift#L1006) following the Swift [SIMD vectors](https://github.com/apple/swift-evolution/blob/master/proposals/0229-simd.md) proposal.

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -13,45 +13,61 @@
 //===------------------------------------------------------------------------------------------===//
 
-public extension Tensor where Scalar: TensorFlowFloatingPoint {
+public extension Tensor where Scalar: BinaryFloatingPoint {
```
We should use `TensorFlowFloatingPoint` here for consistency with other core APIs. You can then drop the `where` clause in `@differentiable` and move the VJP to the same extension.

eaplatanios commented:
@rxwei What should I do about the `#tfop` limitation mentioned above?

rxwei commented:
`#tfop` does support input lists and output lists. It supports anything that conforms to `TensorArrayProtocol`. Just use an array as arguments and/or results.

rxwei commented on Sources/DeepLearning/Loss.swift:
```diff
@@ -13,7 +13,7 @@
 
 #if !COMPILING_TENSORFLOW_MODULE
-import TensorFlow
+@_exported import TensorFlow
```
Only one `@_exported` is needed in all these files.

eaplatanios replied:
What does `@_exported` do?

rxwei replied:
It makes whoever imports `DeepLearning` also import `TensorFlow`.

eaplatanios replied:
Thanks!

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+    /// - Returns: The gathered tensor.
+    @inlinable
+    // @differentiable(vjp: _vjpGathered where Scalar: TensorFlowFloatingPoint)
+    func gathered<I: TensorFlowInteger>(
```
I think "gathering" makes slightly more grammatical sense than "gathered".

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+    /// 
+    /// - Returns: The gathered tensor.
+    @inlinable
+    func batchGathered<I: TensorFlowInteger>(
```
Same. I think "batchGathering" is better.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+        let castedShape = Tensor<I>(shapeTensor)
+        var batchIndices = indices
+        var accumulated = Tensor<I>(ones: [])
+        for d in (1 ... batchDims).reversed() {
```
```suggestion
        for d in (1...batchDims).reversed() {
```

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+        var batchIndices = indices
+        var accumulated = Tensor<I>(ones: [])
+        for d in (1 ... batchDims).reversed() {
+        accumulated *= castedShape[d]
```
Indent the `for` body.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+        let posAxis = axis < 0 ? axis + rank : axis
+        let leadingSize = shapeTensor[posAxis ..< posAxis + mask.rank].product().rankLifted()
+        let reshapedTensor = reshaped(
+        toShape: shapeTensor[..<Int(posAxis)].concatenated(
```
Indent

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+        let leadingSize = shapeTensor[posAxis ..< posAxis + mask.rank].product().rankLifted()
+        let reshapedTensor = reshaped(
+        toShape: shapeTensor[..<Int(posAxis)].concatenated(
+            with: [leadingSize, shapeTensor[Int(posAxis + mask.rank)...]]))
```
Indent further.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+    }
+}
+
+public extension Tensor where Scalar: TensorFlowScalar {
```
`Scalar` always conforms to `TensorFlowScalar`. This constraint is redundant.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+            to: dValue,
+            stride: Tensor<I>(ones: [])
+        ) * accumulated
+        let dShape = Tensor<Int32>(d - 1).packed(with: [
```
Where is `packed` defined?

eaplatanios commented:
> `#tfop` does support input lists and output lists. It supports anything that conforms to `TensorArrayProtocol`. Just use an array as arguments and/or results.

In that case, why does the `Raw` module generation script say it doesn't? Is it just that script that does not support them?

rxwei commented:
> In that case, why does the `Raw` module generation script say it doesn't? Is it just that script that does not support them?

For historical reasons. GPE doesn't support variable-sized inputs/outputs, and `swift-bindings` started to exist when only GPE existed. We never updated the raw op generation script to generate array types, after eager execution became possible.

eaplatanios commented:
In that case, I made a [PR](https://github.com/tensorflow/swift-bindings/pull/21) to add support for them in the `Raw` module generation script.

eaplatanios commented:
I also addressed your comments, but will add the missing VJPs tomorrow because I have to finish up some other stuff tonight.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+    @inlinable
+    // @differentiable(vjp: _vjpGathering where Scalar: TensorFlowFloatingPoint)
+    func gathering<I: TensorFlowInteger>(
+        atIndices indices: Tensor<I>, 
```
Since our index type is `Int32`, this argument should be `Tensor<Int32>` for consistency, instead of being generic. Same for other APIs in this file that take indices.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,303 @@
+    /// - Returns: `(self.rank - K + 1)`-dimensional tensor populated by entries in this tensor 
+    ///   corresponding to `true` values in `mask`.
+    @inlinable
+    func masked(with mask: Tensor<Bool>, alongAxis axis: Int32 = 0) -> Tensor {
```
Although TensorFlow calls it `boolean_mask`, I think it's a bit misleading in that it changes the shape of the tensor. I see it as another form of `gathering`. Seems better to call it `gathering(where:alongAxis:)`.

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -94,12 +96,22 @@ public enum Padding {
     }
+
+    @inlinable
+    var explicitPaddings: [Int32] {
```
This should be `internal` since it's only for internal use.

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -86,6 +86,8 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
 // @_frozen // SR-9739
 public enum Padding {
+    /// The "explicit" padding scheme.
+    case explicit(paddings: [Int32])
```
Remove the `paddings:` label since 1) it's merely repeating the type name, and 2) the plurality is fully conveyed by type information (array).

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -65,45 +79,15 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
 // @_frozen // SR-9739
 public enum Padding {
+    /// The "explicit" padding scheme.
+    case explicit(_ paddings: [Int32])
```
Enum associated value lists are different from function parameter lists. This can just be a type, instead of `_ label:`.

```suggestion
    case explicit([Int32])
```

eaplatanios replied:
That makes sense. I fixed this both here and in the other PR. :)

eaplatanios commented:
@rxwei should I move the advanced indexing PR updates in this PR so that it's part of swift-apis? My PR on the core TF repo should be merged soon and will unblock progress in that direction.

rxwei commented:
I think not, since it’s going to be in conflict with the existing subscript defined in the existing version of stdlib that is used to build this library. Adding up all the different PRs updating checkouts, it’s going to be a mess. Plus, a ton of compiler tests rely on subscript.

eaplatanios commented:
I added all of the missing VJPs and a couple more ops. One issue is, I can't compile this on my machine. For some reason the compiler crashes. @rxwei could you please check if it compiles for you?

eaplatanios commented:
@rxwei @marcrasi I believe there is still an issue with array differentiability crashing. I tried to find where the crash is coming from by commenting out parts of the code (the reported location for errors is not really helpful sometimes). I noticed that the VJPs for the `split` functions seem to be causing a crash with the following error:

```
1.      While emitting SIL for '_vjpSplit(sizes:alongAxis:)' (at /Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:276:5)
2.      While silgen emitFunction SIL function "@$s10TensorFlow0A0V12DeepLearningAA0aB13FloatingPointRzrlE9_vjpSplit5sizes9alongAxisSayACyxGG_AISass14DifferentiableRzlE0L4ViewVyAI_GctACys5Int32VG_SitF".
 for '_vjpSplit(sizes:alongAxis:)' (at /Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:276:5)
3.      While silgen closureexpr SIL function "@$s10TensorFlow0A0V12DeepLearningAA0aB13FloatingPointRzrlE9_vjpSplit5sizes9alongAxisSayACyxGG_AISass14DifferentiableRzlE0L4ViewVyAI_GctACys5Int32VG_SitFAiNcfU_".
 for expression at [/Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:281:25 - line:281:79] RangeText="{ v in Tensor(concatenating: v.base, alongAxis: axis) "
0  swift                    0x000000010cc604c5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x000000010cc5f765 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x000000010cc60ab2 SignalHandler(int) + 258
3  libsystem_platform.dylib 0x00007fff60157b5d _sigtramp + 29
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 2682946752
5  swift                    0x0000000109e9d7e6 swift::SILFunctionBuilder::addFunctionAttributes(swift::SILFunction*, swift::DeclAttributes&, swift::SILModule&, swift::SILDeclRef) + 1606
6  swift                    0x0000000109e9e2b5 swift::SILFunctionBuilder::getOrCreateFunction(swift::SILLocation, swift::SILDeclRef, swift::ForDefinition_t, swift::ProfileCounter) + 1381
7  swift                    0x000000010980dc11 swift::Lowering::SILGenModule::getFunction(swift::SILDeclRef, swift::ForDefinition_t) + 321
8  swift                    0x00000001098f7073 swift::Lowering::SILGenFunction::emitGlobalFunctionRef(swift::SILLocation, swift::SILDeclRef, swift::Lowering::SILConstantInfo, bool) + 515
9  swift                    0x0000000109823968 (anonymous namespace)::Callee::getFnValue(swift::Lowering::SILGenFunction&, bool, llvm::Optional<swift::Lowering::ManagedValue>) const & + 1336
10 swift                    0x000000010982066d (anonymous namespace)::CallEmission::apply(swift::Lowering::SGFContext) + 3405
11 swift                    0x000000010981f81f swift::Lowering::SILGenFunction::emitApplyExpr(swift::Expr*, swift::Lowering::SGFContext) + 2111
12 swift                    0x000000010988b9e0 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 80
13 swift                    0x000000010987d340 swift::Lowering::SILGenFunction::emitRValue(swift::Expr*, swift::Lowering::SGFContext) + 32
14 swift                    0x00000001098f360f swift::Lowering::SILGenFunction::emitReturnExpr(swift::SILLocation, swift::Expr*) + 383
15 swift                    0x00000001098f1901 swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 8465
16 swift                    0x00000001098f179b swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 8107
17 swift                    0x00000001098ef7e5 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
18 swift                    0x00000001098a4bde swift::Lowering::SILGenFunction::emitClosure(swift::AbstractClosureExpr*) + 286
19 swift                    0x00000001098101aa swift::Lowering::SILGenModule::emitClosure(swift::AbstractClosureExpr*) + 330
20 swift                    0x0000000109898ddc (anonymous namespace)::RValueEmitter::visitAbstractClosureExpr(swift::AbstractClosureExpr*, swift::Lowering::SGFContext) + 44
21 swift                    0x000000010988ba00 swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 112
22 swift                    0x0000000109891f5b swift::ASTVisitor<(anonymous namespace)::RValueEmitter, swift::Lowering::RValue, void, void, void, void, void, swift::Lowering::SGFContext>::visit(swift::Expr*, swift::Lowering::SGFContext) + 26059
23 swift                    0x000000010987d340 swift::Lowering::SILGenFunction::emitRValue(swift::Expr*, swift::Lowering::SGFContext) + 32
24 swift                    0x00000001098f360f swift::Lowering::SILGenFunction::emitReturnExpr(swift::SILLocation, swift::Expr*) + 383
25 swift                    0x00000001098f1901 swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 8465
26 swift                    0x00000001098f179b swift::ASTVisitor<(anonymous namespace)::StmtEmitter, void, void, void, void, void, void>::visit(swift::Stmt*) + 8107
27 swift                    0x00000001098ef7e5 swift::Lowering::SILGenFunction::emitStmt(swift::Stmt*) + 21
28 swift                    0x00000001098a43ca swift::Lowering::SILGenFunction::emitFunction(swift::FuncDecl*) + 426
29 swift                    0x000000010981840c swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*)::$_1::operator()(swift::SILFunction*) const + 268
30 swift                    0x000000010980e48e swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*) + 734
31 swift                    0x00000001098fcad9 swift::ASTVisitor<SILGenExtension, void, void, void, void, void, void>::visit(swift::Decl*) + 89
32 swift                    0x00000001098fa6eb SILGenExtension::emitExtension(swift::ExtensionDecl*) + 59
33 swift                    0x00000001098fa6a5 swift::Lowering::SILGenModule::visitExtensionDecl(swift::ExtensionDecl*) + 21
34 swift                    0x00000001098146a6 swift::Lowering::SILGenModule::emitSourceFile(swift::SourceFile*) + 838
35 swift                    0x00000001098155c5 swift::SILModule::constructSIL(swift::ModuleDecl*, swift::SILOptions&, swift::FileUnit*) + 293
36 swift                    0x0000000109815b59 swift::performSILGeneration(swift::FileUnit&, swift::SILOptions&) + 41
37 swift                    0x0000000108f9650e performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 7934
38 swift                    0x0000000108f935a8 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3016
39 swift                    0x0000000108f452d2 main + 690
40 libdyld.dylib            0x00007fff5ff723d5 start + 1
41 libdyld.dylib            0x0000000000000033 start + 2684935263
```

If I uncomment other parts of the code I get the same error but in different places. Do you have any ide of what may be causing this?

rxwei commented:
I'll take a look later today!

eaplatanios commented:
@rxwei I just managed to make it compile with lots of random fixes. Error reporting is currently pretty bad making hard to debug. A few of the changes are:

- `Tensor.init(concatenating:)` is defined in the stdlib so commenting it out helps. Note that the error does not say that there is an ambiguous function call, but we rather just get a crash like the one listed above.
- Remove `@differentiable` from `batchGathering` and from the boolean mask-based `gathering`. Note that again, the error is not that the functions cannot be differentiated, but we rather just get a crash like the one listed above.
- Converted a `Tensor<Int32>([someInt32Tensor])` to `someInt32Tensor.rankLifted()`. Not sure why this was wrong, but again the error reported was just the above crash log.
- Converted some `Tensor<Int32>(0 ..< Int32(4))`, for example, to `Tensor<Int32>(rangeFrom: 0, to: Int32(4), stride: 1)`.

I pushed the above changes so this should be fine to merge now. I'll also make a paired PR on stdlib, removing `Tensor.init(concatenating:)` from there.

eaplatanios commented:
Created the paired [PR #24161 for the swift stdlib](https://github.com/apple/swift/pull/24161).

eaplatanios commented:
I also added an `Optimizable` protocol to make all implemented optimizers operate on more than just layers.

eaplatanios commented:
I updated the PR description to list all the changes this PR introduces.

rxwei commented:
Thanks! Since the main apple/swift build needs to pull tensorflow/swift-apis and use its files, changing the directory structure may require changes in build-script. It's better to revisit this next week once we are done with the v0.3 release.

eaplatanios commented:
Sounds good! FYI it seems to compile fine on my machine for now, but if I run into any issues I'll report them here. I'll keep working on this until we merge it after the v0.3 release.

rxwei commented:
Hi @eaplatanios, if it's not too much trouble, would you mind separating the new operators out of this PR and make this PR be just about restructuring? In that case we don't have to wait till the next week to merge it.

eaplatanios commented:
@rxwei sure I can do that. In that case, should I also move over more ops from stdlib to this repo? I was planning to do that anyway, but this way we can make this PR about moving stuff over from stdlib.

rxwei commented:
Sounds good!

eaplatanios commented:
@rxwei I'm about to start moving over ops but I was thinking that it might be easier if I just comment out the new ops and VJPs and make a new PR later on uncommenting them and making any necessary changes. It's definitely not pretty but it would be more efficient and would not affect the v0.3 release. Would that be ok? If not I can always completely remove them for now.

eaplatanios commented:
I've also found some bugs in the ops I move over (for example, `softmax(_:alongAxis:)` has a bug). Is it ok if I also fix some minor bugs like that?

eaplatanios commented:
@rxwei I removed all the newly added ops and VJPs except for the `Tensor(stacking:alongAxis:)` constructor (since it is a slight generalization of one that was already there and tested) and the VJPs for the `Tensor(stacking:alongAxis:)` and `Tensor(concatenating:alongAxis:)` initializers since these were not there, but should be used very often in practice and thus would be nice to include soon. There are the only new functions added. I'll keep moving over ops from stdlib. My convention is to move pretty much everything that does not depend on internal functionality (such as some of the tensor constructors). You can see that in the current work-in-progress. I should be done pretty soon though.

rxwei commented:
Very impressive progress! Thanks for taking a lead on this huge TODO on our list :)

> My convention is to move pretty much everything that does not depend on internal functionality (such as some of the tensor constructors). You can see that in the current work-in-progress. I should be done pretty soon though.

Sounds great. It's absolutely the right thing to do!

eaplatanios commented:
Thanks Richard! :) I think most things have already been moved over. I'm compiling now to see if everything works fine. I was also thinking of moving the `PythonConversion.swift` contents over as they're also basically a high-level API.

To that end, it may also be meaningful to rename the main library in `swift-apis` to something other than `DeepLearning` at some point, as it's not really specific to deep learning in any way.

rxwei commented:
It would also be useful and important to _copy_ the API tests (`test/TensorFlowRuntime/tensor.swift` and `test/TensorFlowRuntime/tensor_api.swift`) over here and adapt them to the XCTest test suite. The reason we want to do this is that if we happen to break tests in tensorflow/swift-apis, we should be able to find out before doing an update-checkout PR on apple/swift. The reason we want to copy them instead of moving them over is that apple/swift has an effective CI machine with GPU. Once we set up GPU CI here, we can remove them.

rxwei commented:
> To that end, it may also be meaningful to rename the main library in `swift-apis` to something other than `DeepLearning` at some point, as it's not really specific to deep learning in any way.

Agreed. That can definitely be done in a future PR after v0.3. Maybe call it `Numerics`?

eaplatanios commented:
@rxwei I've been moving tests and will push the changes soon but was wondering how to deal with the `TPU` and `CUDA` flags, and also with the `testAllBackends` helper.

rxwei commented:
You don't need to move CUDA-related flags or the `testAllBackends` helper. Those test cases should become methods in XCTest.

eaplatanios commented:
So, I don't need to do anything about this part in the end of `tensor.swift`?

```
// For now it is sufficient to run remote tests with test cases in this file
// only. When creating new test files, consider simply calling runAllTests().
#if CUDA
// RemoteSession does not work for GPU because partitioning logic gets confused
// with multiple devices.
runAllTests()
#else
runAllTestsWithRemoteSession()
#endif  // CUDA
```

rxwei commented:
Nope.

eaplatanios commented:
@rxwei I brought in the tests from `tensor.swift` and `tensor_api.swift` and also the contents of `PythonConversion.swift`. These changes are also reflected in the [corresponding stdlib PR](https://github.com/apple/swift/pull/24161) (although the title is not so relevant anymore :)).

rxwei commented:
Thanks! I actually meant we should keep those tests in stdlib in order to run GPU tests in the near term. Could you add those test files back in stdlib?

eaplatanios commented:
Oops sorry, I missed that. So you mean keep them in stdlib and also keep their copies in swift-apis that I made?

rxwei commented:
Yep! We can remove the stdlib ones when GPU CI is set up for swift-apis.

eaplatanios commented:
Done. :)

rxwei commented:
Awesome. Is it ready to review?

eaplatanios commented:
Yes, I believe so.

rxwei commented on Sources/DeepLearning/Initializers.swift:
```diff
@@ -13,9 +13,295 @@
+// Casting
+//===------------------------------------------------------------------------------------------===//
+
+public extension Tensor where Scalar : Numeric {
```
In the swift-apis library we are no longer using the stdlib code styles but the Google Swift Style Guide instead (except for 4-space indentation). One difference is that there's no space before any colon, including type inheritance clauses. 
```suggestion
public extension Tensor where Scalar: Numeric {
```

rxwei commented on Sources/DeepLearning/Initializers.swift:
````diff
@@ -13,9 +13,295 @@
+    /// Tensor(stacking: [x, y, z], alongAxis: 1) // is [[1, 2, 3], [4, 5, 6]]
+    /// ```
+    ///
+    /// This is the opposite of `Tensor.unstacked`.
````
Thanks a lot for the detailed doc comments! Since this comment mentions `unstacked`, would you like to include it in this PR as well? Also, what's the type signature of it?

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -128,7 +138,7 @@ public extension Padding {
 }
 
-public extension Tensor where Scalar: TensorFlowFloatingPoint {
+public extension Tensor where Scalar : TensorFlowFloatingPoint {
```
```suggestion
public extension Tensor where Scalar: TensorFlowFloatingPoint {
```

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -0,0 +1,1508 @@
+    /// A scalar zero tensor.
+    @inlinable
+    public static var zero: Tensor {
+        get {
```
Omit `get {}` when there isn't a `set {}`.

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -0,0 +1,1508 @@
+// Additive Group
+//===------------------------------------------------------------------------------------------===//
+
+extension Tensor : AdditiveArithmetic where Scalar : Numeric {
```
```suggestion
extension Tensor: AdditiveArithmetic where Scalar: Numeric {
```

rxwei commented on Sources/DeepLearning/Operators/README.md:
```diff
@@ -0,0 +1,41 @@
+# Ops and Convenience Methods
```
```suggestion
# Operators and Convenience Methods
```

rxwei commented on Sources/DeepLearning/Operators/README.md:
```diff
@@ -0,0 +1,41 @@
+Inputs and outputs must be of TensorHandle, ResourceHandle, or VariantHandle
+type. These are magic types known to the compiler.
+
+## Auto-Differentiation Support
```
```suggestion
## Automatic Differentiation Support
```

rxwei commented on Sources/DeepLearning/Optimizer.swift:
```diff
@@ -16,13 +16,16 @@
 import TensorFlow
 #endif
 
+public protocol Optimizable: Differentiable & KeyPathIterable
```
Thanks for introducing this change! However, empty protocols that are only used for joining constraints are generally discouraged in Swift. Could you revert it so that we won't introduce any API breakage in this PR?

eaplatanios replied:
Sounds good! Is it ok if I keep the change of `Model: Layer` -> `Model: Differentiable & KeyPathIterable` (with the where clause) for the optimizer implementations in the file? This is so they can be used for more than just layers.

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -172,14 +182,13 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
-                v.convolved2D(withFilter: filter, strides: strides, padding: padding)
-            )
+            (
+                self.conv2DBackpropFilter(
```
Indent by 4 from `return`.

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -0,0 +1,1508 @@
+    /// Returns the quotient of dividing the first tensor by the second.
+    /// - Note: `/` supports broadcasting.
+    @inlinable
+    @differentiable(vjp: _vjpDivide(lhs:rhs:) where Scalar : TensorFlowFloatingPoint)
```
```suggestion
    @differentiable(vjp: _vjpDivide(lhs:rhs:) where Scalar: TensorFlowFloatingPoint)
```

rxwei commented on Sources/DeepLearning/Tensors.swift:
```diff
@@ -0,0 +1,127 @@
+  /// The rank of the tensor, represented as a `Tensor<Int32>`.
+  @inlinable
+  var rankTensor: Tensor<Int32> {
+    get {
```
Omit `get {}` when there isn't a `set {}`.

rxwei commented on Sources/DeepLearning/Gradients.swift:
```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Could you rename this file to `DifferentialOperators.swift`?

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -0,0 +1,1509 @@
+    }
+
+    @inlinable
+    static func *= (lhs: inout Tensor, rhs: Scalar) {
```
Could you add a doc comment here?

rxwei commented on Sources/DeepLearning/Operators/README.md:
```diff
@@ -0,0 +1,41 @@
+we also define some helper function wrappers, e.g. to make things symmetric
+and generally feel nice to use.
+
+The ops themselves are defined by the primitive `#tfop(...)` syntax, here 
```
Most contents in this README file are actually out of date, since the compiler doesn't do GPE, there's no use of `#tfop`, and the `vjp:` label in `@differentiable` is considered an internal interface. The public interface is `@differentiating` specified on VJP functions. I think removing this file makes most sense. WDYT?

eaplatanios replied:
Sounds good! I actually wasn't sure what the status is with respect to some of these comments and just wanted to make sure I don't delete them before you take a look. Btw, I think there's also some random TODOs throughout the code that may be outdated, but I haven't removed any of them for now.

rxwei replied:
Thanks a lot for keeping things around!

eaplatanios commented:
@rxwei Thanks for the feedback! I addressed all the comments. I had to add support for the `Unpack` op in the `ArrayOps.swift` file in the stdlib in order to support `Tensor.unstack`. Also, I removed the `Optimizable` protocol, as per your suggestion, but changed `Model: Layer` to `Model: Differentiable & KeyPathIterable` (with the where clause on `AllDifferentiableVariables`) for the optimizer implementations in the file, so that they can be used for more than just layers.

eaplatanios commented:
I believe that adding the VJPs for the stacking and the concatenating initializers to this PR would be good because they appear pretty frequently. I have them implemented so I could add them if we decide to do so.

rxwei commented on Sources/DeepLearning/Optimizer.swift:
```diff
@@ -45,8 +45,9 @@ fileprivate extension Tensor where Scalar: Numeric {
 /// https://arxiv.org/abs/1412.6980v8)
-public class Adam<Model: Layer>: Optimizer
-    where Model.AllDifferentiableVariables == Model.CotangentVector {
+public class Adam<Model: Differentiable & KeyPathIterable>: Optimizer
```
This still has a chance to break existing notebooks are we are preparing with fast.ai.  I'd suggest not introducing any functional changes to existing abstractions until we are ready to do so.

eaplatanios replied:
Oh I see. Ok I reverted that.

rxwei replied:
Yeah, we are in a pre-rerelease phase and the fast.ai course is next week. Thanks for understanding! Let's revisit this in a few weeks.

rxwei commented:
> I believe that adding the VJPs for the stacking and the concatenating initializers to this PR would be good because they appear pretty frequently. I have them implemented so I could add them if we decide to do so.

That sounds good.

rxwei commented:
We just received a small patch on apple/swift that changes the `expandingShape(at:)` operator (https://github.com/apple/swift/pull/24191). Could you incorporate that once that gets merged?

eaplatanios commented:
Yes definitely. I'll add them once merged. I already added their previous fix for the VJPs of mean and sum. I also just pushed edits that add VJPs for `Tensor.init(stacking:alongAxis)` and  `Tensor.init(concatenating:alongAxis)`.

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -1341,9 +1341,7 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
-	        for i in axes.array.scalars {
-                result = result.expandingShape(at: Int(i))
-            }
+	        for i in axes.array.scalars { result = result.expandingShape(at: Int(i)) }
```
Looks like tabs are used instead of spaces.

rxwei commented on Sources/DeepLearning/Operators/Math.swift:
```diff
@@ -1361,9 +1359,7 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
-	        for i in axes.array.scalars {
-                result = result.expandingShape(at: Int(i))
-            }
+	        for i in axes.array.scalars { result = result.expandingShape(at: Int(i)) }
```
Same here.

eaplatanios commented:
I also just added the changes from the merged PR (apple/swift#24191) in stdlib.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -220,9 +230,9 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
-        let value = expandingShape(at: shapeIndex)
-        return (value, { v in v.squeezingShape(at: shapeIndex) })
+    func _vjpExpandingShape(at axes: [Int]) -> (Tensor, (Tensor) -> Tensor) {
+	    let value = self.expandingShape(at: axes)
```
same here.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -154,12 +154,22 @@ public extension Tensor {
-    func expandingShape(at shapeIndex: Int) -> Tensor {
-        return Raw.expandDims(self, dim: Tensor<Int32>(Int32(shapeIndex)))
+    func expandingShape(at axes: [Int]) -> Tensor {
+	    var result = self
```
tabs -> spaces.

rxwei commented on Tests/DeepLearningTests/TensorTests.swift:
```diff
@@ -0,0 +1,81 @@
+        XCTAssertEqual(0, selectValue(true).scalar)
+    }
+
+    @inline(never)
```
Could you help remove `@inline(never)` from all of these files? It's a GPE artifact that is no longer required.

rxwei commented on Tests/DeepLearningTests/TensorTests.swift:
```diff
@@ -0,0 +1,81 @@
+        let ones = Tensor<Int32>(ones: [1, 2, 2, 2, 2, 2, 1])
+        XCTAssertEqual([1, 2, 2, 2, 2, 2, 1], ones.shape)
+    }
+}
```
In order to make tests run on Linux, could you add a `allTests` computed property like [this one](https://github.com/tensorflow/swift-apis/blob/3f716844e38b462cf6e6c0ba70b56b67fa5cf9fb/Tests/DeepLearningTests/LayerTests.swift#L123) in every test file?

rxwei commented on Sources/DeepLearning/Operators/NN.swift:
```diff
@@ -16,22 +16,36 @@
 public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Computes the batch normalized tensor along the specified axis.
+    ///
+    /// Specifically, returns `(self - mu) / (var + epsilon) * gamma + beta` where `mu` and `var` are 
```
Is this line over 100 columns?

rxwei commented on Sources/DeepLearning/PythonConversion.swift:
```diff
@@ -0,0 +1,174 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Could you please apply 4-space indentation in this file?

rxwei commented on Sources/DeepLearning/Random.swift:
```diff
@@ -429,8 +429,8 @@ public struct UniformIntegerDistribution<T: FixedWidthInteger>: RandomDistributi
-public struct UniformFloatingPointDistribution<T : BinaryFloatingPoint>: RandomDistribution
-  where T.RawSignificand : FixedWidthInteger {
+public struct UniformFloatingPointDistribution<T: BinaryFloatingPoint>: RandomDistribution
+  where T.RawSignificand: FixedWidthInteger {
```
Could you help apply 4-space indentation in this file as well?

eaplatanios commented:
I addressed all the review comments. Regarding the tests, I'm about to run them with all the last changes right after Swift finishes compiling on my machine.

eaplatanios commented:
@rxwei All tests pass on my machine.

rxwei commented:
Could you update the update-checkout file in the corresponding PR once this gets merged?

eaplatanios commented:
Done in apple/swift#24161 .

rxwei commented:
My apologies, I shouldn't have merged this PR since it did not compile or pass XCTests. We should make sure it compiles and passes tests. @eaplatanios could you look into it?

eaplatanios commented:
This compiles and passes all tests on my machine, when I follow the following process:
- Copy this over to the `tensorflow-swift-apis` directory (same as what the update-checkouts command would do).
- Compile apple/swift, run all tests, and build toolchain.

This process all works fine.

The problem is when I try to compile this on its own using the built toolchain from above. That toolchain already contains all symbols from swift-apis in the stdlib, as it builds swift-apis together with the stdlib. This means that if I try to compile this with that toolchain I get many errors about ambiguous functions.

rxwei commented:
Yeah, I understand the apple/swift path compiles and passes tests. To ensure swift-apis can be developed individually, we'll also need to make sure it works with Swift Package Manager.

Earlier there were already quite a few operators and layers that were both defined in here and distributed as part of the toolchain. There were no ambiguity errors because overriding a symbol in a different module is okay. There may be some subtleties in the newly migrated APIs. It would be useful to see what exactly those ambiguities are.

rxwei commented:
Let me know if you have any question via email or a new PR. I'm more than happy to help with those issues!

eaplatanios commented:
I see, yes that makes sense. So, the output when compiling swift-apis looks as follows:
- Multiple warnings about protocol conformances that were declared in stdlib. This is expected since the "namescope" does not matter in this case.
- Multiple errors about the `@_semantics` attribute I added in the last PR. These look like: `error: '@_semantics' attribute cannot be applied to this declaration`.
- Error in `PythonConversion.swift`: `use of unresolved identifier 'debugLog'`.
- Error that I can't figure out:
```
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:342:48: error: ambiguous use of 'subscript(_:)'
        let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
                                               ^
TensorFlow.Tensor<τ_0_0>:2:23: note: found this candidate
    @inlinable public subscript(ranges: TensorRangeExpression...) -> TensorFlow.Tensor<Scalar> { get set }
                      ^
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:633:17: note: found this candidate
    fileprivate subscript(_ ranges: TensorRangeExpression...) -> Tensor {
                ^
```

eaplatanios commented:
That's all actually so it should be easy to resolve.

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,716 @@
+    /// - Returns: Array containing the tensors parts.
+    @inlinable
+    @differentiable(vjp: _vjpSplit(numSplits:alongAxis:) where Scalar: TensorFlowFloatingPoint)
+    func split(numSplits: Int, alongAxis axis: Int = 0) -> [Tensor] {
```
I missed this newly added operator. Swift APIs usually don't use the "num" prefix to denote count, but use a `Count` suffix instead. In this case, a single `count` may also be fine. You could call this `split(count:alongAxis:)` or follow the standard library precedent: [split(maxSplits:omittingEmptySubsequences:whereSeparator:)](https://developer.apple.com/documentation/swift/array/2294101-split).

rxwei commented on Sources/DeepLearning/Operators/Basic.swift:
```diff
@@ -0,0 +1,716 @@
+    /// - Returns: Array containing the unstacked tensors.
+    @inlinable
+    @differentiable(vjp: _vjpUnstack(alongAxis:) where Scalar: TensorFlowFloatingPoint)
+    func unstack(alongAxis axis: Int = 0) -> [Tensor] {
```
Functions that do not have side effects should be named like nouns (see [Swift API Design Guidelines - Naming - Strive For Fluent Usage](https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage)). Since this method doesn't have side effects, it a gerund (`unstacking(alongAxis:)`) would be a better name.

rxwei commented:
> * Multiple errors about the `@_semantics` attribute I added in the last PR. These look like: `error: '@_semantics' attribute cannot be applied to this declaration`.

I think that `@_semantics` is only available on explicit property getters (`get {}`).

> * Error in `PythonConversion.swift`: `use of unresolved identifier 'debugLog'`.

Feel free to comment out those `debugLog`s for now!



> * Error that I can't figure out:
> 
> ```
> /Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:342:48: error: ambiguous use of 'subscript(_:)'
>         let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
>                                                ^
> TensorFlow.Tensor<τ_0_0>:2:23: note: found this candidate
>     @inlinable public subscript(ranges: TensorRangeExpression...) -> TensorFlow.Tensor<Scalar> { get set }
>                       ^
> /Users/eaplatanios/Development/GitHub/swift-apis/Sources/DeepLearning/Operators/Basic.swift:633:17: note: found this candidate
>     fileprivate subscript(_ ranges: TensorRangeExpression...) -> Tensor {
>                 ^
> ```

I'm not sure either. Why is the one on 633 `fileprivate` in the error message?

eaplatanios commented:
> I think that `@_semantics` is only available on explicit property getters (`get {}`).

That fixes it!

> Feel free to comment out those `debugLog`s for now!

Done.

> I'm not sure either. Why is the one on 633 `fileprivate` in the error message?

This was just something I was trying. Could this error be because the `subscript` function is defined in a tensor extension and that may also cause a conflict when redefined across different modules? `TensorFlow` is imported in that file and so there is already a `Tensor.subscript` with the same signature defined in scope.

rxwei commented:
I don't have a good solution to that. How about defining indexing logic in stdlib for now?

eaplatanios commented:
That's ok for now but I also believe it is a more important issue that'll likely pop up again with other extensions.

eaplatanios commented:
Actually I believe this might happen a lot in the future, given that a lot of the added functionality is in the form of extensions to `Tensor`. Although, I am curious as to why we're only observing this for `subscript` and not for other functions that are also implemented in extensions of `Tensor`.

rxwei commented:
Yeah, I think we should definitely get it resolved sooner rather than later. The Foundation model would be good to follow. Other people on the team have talked about moving TensorFlow out of the compiler repo as well (@pschuh @jekbradbury).

eaplatanios commented:
I would be happy to help with that because even the changes introduced in this PR seem hard to properly handle (in that swift-apis cannot easily be compiled independently using S4TF). And yes, I would lean in favor of moving TensorFlow out of the compiler repo if that is an option. Could you please actually describe the Foundation model at a high level?

eaplatanios commented:
Most importantly actually, I'm curious about how the deal with the ambiguous symbols issue that we are running into.

rxwei commented:
It's really awesome to hear you are interested in helping out! 

Foundation sources are located in [apple/swift-corelibs-foundation](https://github.com/apple/swift-corelibs-foundation), and it's being pulled into the build directory by `update-checkout`. When building using `build-toolchain`, it is being included as one of the core libraries in toolchains so that users can import `Foundation`.

Foundation is a little different in that it's not a SwiftPM package, but I suspect that's not the difficult part.

rxwei commented:
> Most importantly actually, I'm curious about how the deal with the ambiguous symbols issue that we are running into.

Are you referring to the subscript errors? I'd suggest pulling them back to stdlib for now.

eaplatanios commented:
I've started using S4TF for some new research projects and would generally love to make the process of improving it and extending it smoother. I love the work you guys have put in and all that lies ahead! :)

Regarding foundation, that sounds similar to what's happening now with swift-apis. I believe the main challenge (irrespective of using SwiftPM) is in compiling swift-apis as a standalone library using the S4TF compiler. I believe this would be much easier if we move all of the `TensorFlow` library outside of the compiler repo and into swift-apis. I would be willing to give this a shot, although I'm not sure if there is anything internal being used in the `TensorFlow` directory in stdlib that won't compile if I move it out. I could give it a shot as part of #109 and see if things work out.

rxwei commented:
> I've started using S4TF for some new research projects and would generally love to make the process of improving it and extending it smoother. I love the work you guys have put in and all that lies ahead! :)

That's fantastic. We really really appreciate your effort in driving important changes as well as pushing the limits of the incomplete Swift for TensorFlow.

> I believe the main challenge (irrespective of using SwiftPM) is in compiling swift-apis as a standalone library using the S4TF compiler.

Right. This is why we probably shouldn't attempt this right now. In the TensorFlow module there is legacy code w.r.t. GPE, and also new experimental code built by @bgogul and @pschuh et al.

One option is to keep `TensorHandle` (as well as other important handle types) in the compiler repo as `TensorFlowCore`, and all other things in a separate `TensorFlow` repo.

I feel that tensor APIs being in the compiler repo really starts to limit our development. So it's probably good time to get the discussion started. How about setting up a meeting with the team to discuss this? Are you able to video chat some time this or next week?

vguerra commented:
hello @eaplatanios. Quick question on this PR. Description claims the addition of initializers: `Tensor(zerosLike:)` and `Tensor(onesLike:)` but I don't see them, or maybe I am missing something (?):/.

rxwei commented:
Not sure about the new initializers. The ‘Optimizer’ protocol has been reverted, addressing [this review feedback](https://github.com/tensorflow/swift-apis/pull/70#discussion_r277148057).

vguerra commented:
@rxwei thanks.. missed those comments.

eaplatanios commented:
> I feel that tensor APIs being in the compiler repo really starts to limit our development. So it's probably good time to get the discussion started. How about setting up a meeting with the team to discuss this? Are you able to video chat some time this or next week?

Yes, definitely. That would be great! How about sometime on Wednesday?

Also, I like the `TensorFlowCore` suggestion and that would resolve the ambiguous symbols issues too I believe.

eaplatanios commented:
> hello @eaplatanios. Quick question on this PR. Description claims the addition of initializers: `Tensor(zerosLike:)` and `Tensor(onesLike:)` but I don't see them, or maybe I am missing something (?):/.
> 
> Same goes for protocol `Optimizable`.

Sorry, yes we decided to remove the new ops from this PR to make it about moving stuff over from stdlib first. I have all the new ops implemented but will add them in a separate PR, after we sort out #109 .

vguerra commented:
> Sorry, yes we decided to remove the new ops from this PR to make it about moving stuff over from stdlib first. I have all the new ops implemented but will add them in a separate PR, after we sort out #109 .

Got it! .. thanks!

rxwei commented:
> Yes, definitely. That would be great! How about sometime on Wednesday?

Reached out over email regarding the meeting!
